File: android/app/src/test/java/com/timebox/android/data/BlockIdentityTest.kt
package com.timebox.android.data

import java.time.Instant
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class BlockIdentityTest {
    private val linkedTask = LinkedTask(
        id = 7,
        title = "Prepare launch",
        status = TaskStatus.Open,
        taskTypeId = 2,
        archivedAt = null,
        deletedAt = null,
    )

    private val linked = TimeBlock(
        id = 1,
        lane = Lane.Planned,
        taskTypeId = 2,
        taskTypeName = "Deep work",
        taskId = linkedTask.id,
        task = linkedTask,
        note = null,
        plannedBlockId = null,
        startMinute = 600,
        endMinute = 660,
        name = "Outline session",
    )

    @Test
    fun `linked name is primary and Battle Plan Task is secondary`() {
        assertEquals("Outline session", linked.primaryIdentity())
        assertEquals("Prepare launch", linked.secondaryIdentity())

        val unnamed = linked.copy(name = null)
        assertEquals("Prepare launch", unnamed.primaryIdentity())
        assertEquals("Deep work", unnamed.secondaryIdentity())
    }

    @Test
    fun `unlink keeps name and unnamed unlink uses normal fallback`() {
        val namedUnlinked = linked.copy(taskId = null, task = null)
        assertEquals("Outline session", namedUnlinked.primaryIdentity())
        assertEquals("Deep work", namedUnlinked.secondaryIdentity())

        val unnamedUnlinked = namedUnlinked.copy(name = null)
        assertEquals("Deep work", unnamedUnlinked.primaryIdentity())
        assertNull(unnamedUnlinked.secondaryIdentity())
    }

    @Test
    fun `Actual Block follows the same linked identity precedence`() {
        val actual = ActualBlock(
            id = 3,
            taskTypeId = 2,
            taskTypeName = "Deep work",
            taskId = linkedTask.id,
            task = linkedTask,
            note = null,
            plannedBlockId = null,
            startAt = Instant.EPOCH,
            endAt = Instant.EPOCH.plusSeconds(1800),
            name = "Review session",
        )
        assertEquals("Review session", actual.primaryIdentity())
        assertEquals("Prepare launch", actual.secondaryIdentity())
        assertEquals("Prepare launch", actual.copy(name = null).primaryIdentity())
    }
}

File: artifacts/qa-fixes-2026-09-06/baseline-ebf691e/frontend/src/lib/blockIdentity.test.ts
import { describe, expect, it } from 'vitest'
import type { TimeBlock } from './api'
import { blockPrimaryIdentity, blockSecondaryIdentity } from './blockIdentity'

const base: TimeBlock = {
  id: 1,
  lane: 'planned',
  task_type_id: 2,
  task_type: { id: 2, name: 'Deep work', created_at: '', updated_at: '' },
  task_id: 7,
  task: { id: 7, title: 'Prepare launch', status: 'open', task_type_id: 2 },
  name: null,
  note: null,
  start_minute: 600,
  end_minute: 660,
  created_at: '',
  updated_at: '',
}

describe('Block identity', () => {
  it('puts a linked Block Name first and keeps the Battle Plan Task as context', () => {
    const named = { ...base, name: 'Outline session' }
    expect(blockPrimaryIdentity(named)).toBe('Outline session')
    expect(blockSecondaryIdentity(named)).toBe('Prepare launch')
  })

  it('uses the linked task first when unnamed and keeps meaningful Task Type context', () => {
    expect(blockPrimaryIdentity(base)).toBe('Prepare launch')
    expect(blockSecondaryIdentity(base)).toBe('Deep work')
  })

  it('preserves ordinary fallbacks after unlink without copying the former task title', () => {
    const namedUnlinked = { ...base, task_id: null, task: null, name: 'Outline session' }
    expect(blockPrimaryIdentity(namedUnlinked)).toBe('Outline session')
    expect(blockSecondaryIdentity(namedUnlinked)).toBe('Deep work')

    const unnamedUnlinked = { ...namedUnlinked, name: null }
    expect(blockPrimaryIdentity(unnamedUnlinked)).toBe('Deep work')
    expect(blockSecondaryIdentity(unnamedUnlinked)).toBeNull()
  })
})
